# Networks: structure, evolution & processes
**Internet Analytics - Lab 2**

---

**Group:** *E*

**Names:**

* *Enrico Benedettini*
* *Mak Fazlic*
* *Ali Essonni*
* *Youssef Amine*

---

#### Instructions

*This is a template for part 2 of the lab. Clearly write your answers, comments and interpretations in Markodown cells. Don't forget that you can add $\LaTeX$ equations in these cells. Feel free to add or remove any cell.*

*Please properly comment your code. Code readability will be considered for grading. To avoid long cells of codes in the notebook, you can also embed long python functions and classes in a separate module. Don’t forget to hand in your module if that is the case. In multiple exercises, you are required to come up with your own method to solve various problems. Be creative and clearly motivate and explain your methods. Creativity and clarity will be considered for grading.*

---

## 2.2 Network sampling

#### Exercise 2.7: Random walk on the Facebook network

In [5]:
import requests
import random
import numpy as np

In [6]:
def getData(node):
    URL_TEMPLATE = 'http://iccluster050.iccluster.epfl.ch:5050/v1.0/facebook?user={user_id}'
    url = URL_TEMPLATE.format(user_id=node)
    response = requests.get(url)
    return response.json()

def walker(starting_node, number_nodes):
    i = 0
    node = starting_node
    ages = []
    while i <number_nodes:
        data = getData(node)
        ages.append(data["age"])
        neighboors = data["friends"]
        id = random.randrange(0, len(neighboors))
        node = neighboors[id]
        i = i + 1
    return np.sum(ages) / number_nodes

In [7]:
user_id = "a5771bce93e200c36f7cd9dfd0e5deaa"
users = 1000
avg = walker(user_id,users)

In [8]:
print(f"The average age of facebook users using the user {user_id} as a starting node and using {users} is {avg} years")

The average age of facebook users using the user a5771bce93e200c36f7cd9dfd0e5deaa as a starting node and using 1000 is 25.426 years


#### Exercise 2.8

1. We got 25.4 years as average years which  is not close to the true average which is 45 by 19.6 years
2. The discrepency can be explained by the fact if someone with age x it's more likely that they have friends of the same age so doing the usual average won't work

3. We need to find the unbaised estimator which is by definition :
$$\hat{F} = \frac{\sum_t{f(X_t)/ d_{X_t}}}{\sum_t{1/ d_{X_t}}}$$

In [9]:
#3.
def walker_upgraded(starting_node, number_nodes):
    ages = []
    degrees = []
    n = number_nodes
    node = starting_node
    while n > 0:
        data = getData(node)
        neighboors = data["friends"]
        age = data["age"]
        deg = len(neighboors)
        id = random.randrange(0, len(neighboors))
        node = neighboors[id]
        n = n - 1
        ages.append(age)
        degrees.append(deg)

    ages = np.array(ages)
    degrees = np.array(degrees)
    return np.sum(ages / degrees) / np.sum(1 / degrees)

Now we will try multiple values of users until we find the optimal value of users to get the optimal age average

In [16]:
user_id = "a5771bce93e200c36f7cd9dfd0e5deaa"
users = 10 ** 4
avg = walker_upgraded(user_id, users)

In [18]:
print(f"The average age of poeple on facebook using {users} users is : {avg} years")

The average age of poeple on facebook using 10000 users is : 44.84443081085225 years


After trying multiple values of users we find that our age average (44.8 years) is the closest to the real average of facebook users (45 years) is reached with 10000 users